src/box/watcher.c

```diff
@@ -352,7 +352,7 @@ watchable_worker_f(va_list ap)
 		fiber_check_gc();
 		if (!watchable_run(watchable)) {
 			/* No more watchers to run, wait... */
-			fiber_sleep(TIMEOUT_INFINITY);
+			fiber_yield();
```

Comment:
I don't understand what happens here. Both `fiber_sleep` and `fiber_yield` can be interrupted with `fiber_wakeup`. The only difference is that the sleep will wakeup on its own when the time expires which in case of this timeout would take quite a long time. Why is this change needed?

Reply:
There was a rather strange problem when using `fiber_sleep`. [I just added a call to `box_register_watcher`, after which the tarantool entered an infinite loop.](https://github.com/tarantool/tarantool/issues/9632)
`box_register_watcher` calls `watchable_wakeup_worker` which creates a worker fiber and immediately calls `fiber_wakeup` on it. However, the scheduler wakes up only after the first measurement of the number of events in `tarantool_lua_run_script`.
https://github.com/tarantool/tarantool/blob/ebb170cb8cf2b9c4634bcf0178665909f578c335/src/main.cc#L1058-L1065
After this, the `watchable->worker` wakes up and calls `fiber_sleep` which increases the number of events (`fiber_sleep` -> `fiber_yield_timeout` -> `ev_timer_start` -> `ev_start` -> `ev_ref`). As a result, we get that the number of events has increased after calling `tarantool_lua_run_script`.
https://github.com/tarantool/tarantool/blob/ebb170cb8cf2b9c4634bcf0178665909f578c335/src/main.cc#L1073-L1074
In summary, answering the question what is the difference between `fiber_yield` and `fiber_sleep` - the main difference in our context is that `fiber_sleep` generates a new event, while `fiber_yeild` does not.

Reply:
That seems like a bug in the wakeup code or any other code which relies on this difference. Please, file a separate ticket with this bug extracted into a simple example (AFAIU, you create a fiber which calls sleep and then you do wakeup on it and it doesn't work?).

Reply:
This ticket has already been created #9632. There are several examples there on how to reproduce this.